**Re-review: LGTM** — all changes-requested items addressed (follow-up to #4748639335).

Verified on head `543c171` (2 new commits: `3c5f49c8`, `543c1719`).

---

**What changed since the prior review**

- `3c5f49c8` — All 8 provider files updated: `openai.py`, `azure_openai.py`, `anthropic.py`, `deepseek.py`, `minimax.py`, `lmstudio.py`, `vllm.py`, `ollama.py` — each now passes `config.http_client_proxies` (the raw proxy string/dict) instead of `config.http_client` (the already-built `httpx.Client`) to the provider-specific config constructor.
- `543c1719` — `tests/llms/test_openai.py::test_openai_llm_preserves_proxies_from_base_config` added: constructs `OpenAILLM` directly with a `BaseLlmConfig(http_client_proxies=...)` and asserts the raw value and the built `httpx.Client` are both preserved through to the provider config.
- `tests/llms/test_azure_openai.py::test_generate_with_http_proxies` assertion updated from `proxies=` to `proxy=` (httpx 0.28 API).

---

**Red-green proof — direct-path test (`test_openai_llm_preserves_proxies_from_base_config`)**

GREEN (fix in): `1 passed in 0.81s`

RED (reverted `openai.py` + `base.py` to `HEAD~2`, test kept):
```
FAILED tests/llms/test_openai.py::test_openai_llm_preserves_proxies_from_base_config
AttributeError: 'Client' object has no attribute 'url'
1 failed in 1.49s
```
(The old code passes the `httpx.Client` object where a URL string is expected — exactly the bug.)

GREEN (fix restored): `1 passed in 1.26s`

**Original 7 regression tests:** 7 passed in 3.76s

**CI:** Python SDK builds pass on 3.10, 3.11, 3.12. CI Gate green. Only Vercel failed (advisory — fork permissions, expected on all PRs, not merge-blocking).

**Lint:** `ruff check` on all 13 changed files — `All checks passed!`

---

**Prior minor notes** (not blockers, unchanged)

- `_build_http_client` is still duplicated in `base.py` for LLMs and embeddings — minor; no correctness issue today.
- Test file lives at root `tests/` rather than `tests/configs/` — minor.

Both remain non-blocking suggestions for a later cleanup PR.

---

Ready to squash-merge. (`gh pr merge --squash`)

--- a/mem0/configs/embeddings/base.py
+++ b/mem0/configs/embeddings/base.py
@@ -2,24 +2,8 @@ import os
 from abc import ABC
 from typing import Dict, Optional, Union
 
-import httpx
-
 from mem0.configs.base import AzureConfig
-
-
-def _build_http_client(http_client_proxies: Optional[Union[Dict, str]]) -> Optional[httpx.Client]:
-    """Build an httpx.Client for the given proxy settings.
-
-    httpx >= 0.28 removed the ``proxies=`` argument: a single proxy is passed as
-    ``proxy=`` and per-scheme proxies via ``mounts=``.
-    """
-    if not http_client_proxies:
-        return None
-    if isinstance(http_client_proxies, dict):
-        return httpx.Client(
-            mounts={scheme: httpx.HTTPTransport(proxy=url) for scheme, url in http_client_proxies.items()}
-        )
-    return httpx.Client(proxy=http_client_proxies)
+from mem0.utils.http import build_http_client
 
 
 class BaseEmbedderConfig(ABC):
@@ -97,7 +81,7 @@ class BaseEmbedderConfig(ABC):
 
         # AzureOpenAI specific
         self.http_client_proxies = http_client_proxies
-        self.http_client = _build_http_client(http_client_proxies)
+        self.http_client = build_http_client(http_client_proxies)
 
         # Ollama specific
         self.ollama_base_url = ollama_base_url
@@ -125,4 +109,3 @@ class BaseEmbedderConfig(ABC):
         self.aws_secret_access_key = aws_secret_access_key
         self.aws_session_token = aws_session_token
         self.aws_region = aws_region or os.environ.get("AWS_REGION") or "us-west-2"
-

--- a/mem0/configs/llms/base.py
+++ b/mem0/configs/llms/base.py
@@ -1,22 +1,7 @@
 from abc import ABC
 from typing import Dict, Optional, Union
 
-import httpx
-
-
-def _build_http_client(http_client_proxies: Optional[Union[Dict, str]]) -> Optional[httpx.Client]:
-    """Build an httpx.Client for the given proxy settings.
-
-    httpx >= 0.28 removed the ``proxies=`` argument: a single proxy is passed as
-    ``proxy=`` and per-scheme proxies via ``mounts=``.
-    """
-    if not http_client_proxies:
-        return None
-    if isinstance(http_client_proxies, dict):
-        return httpx.Client(
-            mounts={scheme: httpx.HTTPTransport(proxy=url) for scheme, url in http_client_proxies.items()}
-        )
-    return httpx.Client(proxy=http_client_proxies)
+from mem0.utils.http import build_http_client
 
 
 class BaseLlmConfig(ABC):
@@ -89,7 +74,5 @@ class BaseLlmConfig(ABC):
         self.vision_details = vision_details
         self.reasoning_effort = reasoning_effort
         self.is_reasoning_model = is_reasoning_model
-        # Keep the raw proxies value so it survives config conversion (e.g. in
-        # LlmFactory) instead of only the built client.
         self.http_client_proxies = http_client_proxies
-        self.http_client = _build_http_client(http_client_proxies)
+        self.http_client = build_http_client(http_client_proxies)

--- a/mem0/utils/http.py
+++ b/mem0/utils/http.py
@@ -1,0 +1,13 @@
+from typing import Dict, Optional, Union
+
+import httpx
+
+
+def build_http_client(http_client_proxies: Optional[Union[Dict, str]]) -> Optional[httpx.Client]:
+    if not http_client_proxies:
+        return None
+    if isinstance(http_client_proxies, dict):
+        return httpx.Client(
+            mounts={scheme: httpx.HTTPTransport(proxy=url) for scheme, url in http_client_proxies.items()}
+        )
+    return httpx.Client(proxy=http_client_proxies)

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -17,6 +17,7 @@ dependencies = [
     "qdrant-client>=1.12.0",
     "pydantic>=2.7.3",
     "openai>=1.90.0",
+    "httpx>=0.28.0",
     "posthog>=7.14.0",
     "pytz>=2024.1",
     "sqlalchemy>=2.0.31",

--- a/tests/llms/test_azure_openai.py
+++ b/tests/llms/test_azure_openai.py
@@ -289,8 +289,6 @@ def test_generate_with_http_proxies(default_headers):
             api_version=None,
             default_headers=default_headers,
         )
-        # httpx >= 0.28 removed the ``proxies=`` argument; a single proxy is now
-        # passed as ``proxy=``. _build_http_client builds the client accordingly.
         mock_http_client.assert_called_once_with(proxy="http://testproxy.mem0.net:8000")
 
 

--- a/tests/llms/test_openai.py
+++ b/tests/llms/test_openai.py
@@ -456,10 +456,6 @@ def test_callback_with_tools(mock_openai_client):
 
 
 def test_openai_llm_preserves_proxies_from_base_config(mock_openai_client):
-    # Constructing a provider directly with a BaseLlmConfig (not OpenAIConfig)
-    # must forward the raw proxies value, not the already-built httpx.Client.
-    # The factory path is covered separately; this exercises the direct path
-    # that integration examples and tests/llms/test_openai.py rely on.
     config = BaseLlmConfig(
         model="gpt-4.1-nano-2025-04-14",
         api_key="api_key",

--- a/tests/test_http_client_proxies.py
+++ b/tests/test_http_client_proxies.py
@@ -1,13 +1,3 @@
-"""Regression tests for HTTP proxy support in LLM/embedding configs.
-
-Covers two bugs:
-1. `httpx.Client(proxies=...)` was removed in httpx >= 0.28, so building any
-   config with a proxy raised TypeError.
-2. `LlmFactory` passed the already-built `http_client` where the raw
-   `http_client_proxies` value was expected when converting a BaseLlmConfig to a
-   provider-specific config, silently dropping the user's proxy setting.
-"""
-
 import httpx
 import pytest
 
@@ -20,7 +10,6 @@ from mem0.utils.factory import LlmFactory
 def test_config_with_string_proxy_builds_client(config_cls):
     config = config_cls(http_client_proxies="http://proxy.local:8080")
     assert isinstance(config.http_client, httpx.Client)
-    # The raw value is preserved (not only the built client).
     assert config.http_client_proxies == "http://proxy.local:8080"
 
 
@@ -40,8 +29,6 @@ def test_config_without_proxy_has_no_client(config_cls):
 
 
 def test_llm_factory_preserves_http_client_proxies():
-    # Converting a BaseLlmConfig to a provider config must pass the raw proxies
-    # value through, not the constructed httpx.Client object.
     base = BaseLlmConfig(
         model="gpt-4o-mini",
         api_key="sk-test",

